cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -1466,133 +1327,236 @@ public void persistSubscriberState(String topic, String subscriber, String messa
+
+  private Range getNamespaceAndApplicationRange(String namespaceId, String applicationId) {
+    return Range.singleton(
+      ImmutableList.of(
+        Fields.stringField(StoreDefinition.AppMetadataStore.NAMESPACE_FIELD, namespaceId),
+        Fields.stringField(StoreDefinition.AppMetadataStore.APPLICATION_FIELD, applicationId)));
+  }
+
+  private void writeApplicationSerialized(String namespaceId, String appId, String versionId, String serialized)
+    throws IOException {
+    List<Field<?>> fields = getApplicationPrimaryKeys(namespaceId, appId, versionId);
+    fields.add(Fields.stringField(StoreDefinition.AppMetadataStore.APPLICATION_DATA_FIELD, serialized));
+    applicationSpecificationTable.upsert(fields);
+  }
+
+  private List<Field<?>> getWorkflowPrimaryKeysWithoutNode(ProgramRunId programRunId) {
+    List<Field<?>> fields = new ArrayList<>();
+    fields.add(Fields.stringField(StoreDefinition.AppMetadataStore.NAMESPACE_FIELD, programRunId.getNamespace()));
+    fields.add(Fields.stringField(StoreDefinition.AppMetadataStore.APPLICATION_FIELD, programRunId.getApplication()));
+    fields.add(Fields.stringField(StoreDefinition.AppMetadataStore.VERSION_FIELD, programRunId.getVersion()));
+    fields.add(Fields.stringField(StoreDefinition.AppMetadataStore.PROGRAM_TYPE_FIELD, programRunId.getType().name()));
+    fields.add(Fields.stringField(StoreDefinition.AppMetadataStore.PROGRAM_FIELD, programRunId.getProgram()));
+    fields.add(Fields.stringField(StoreDefinition.AppMetadataStore.RUN_FIELD, programRunId.getRun()));
+    return fields;
+  }
+
+  private List<Field<?>> getWorkflowPrimaryKeys(ProgramRunId programRunId, String nodeId) {
+    List<Field<?>> fields = getWorkflowPrimaryKeysWithoutNode(programRunId);
+    fields.add(Fields.stringField(StoreDefinition.AppMetadataStore.NODE_ID, nodeId));
+    return fields;
+  }
+
+  private List<Field<?>> getCountTypePrefix(String countType) {
+    List<Field<?>> fields = new ArrayList<>();
+    fields.add(Fields.stringField(StoreDefinition.AppMetadataStore.COUNT_TYPE, countType));
+    return fields;
+  }
+
+  private List<Field<?>> getCountNamespacePrefix(String countType, String namespace) {
+    List<Field<?>> fields = getCountTypePrefix(countType);
+    fields.add(Fields.stringField(StoreDefinition.AppMetadataStore.NAMESPACE_FIELD, namespace));
+    return fields;
+  }
+
+  private List<Field<?>> getCountApplicationPrefix(String countType, ApplicationId applicationId) {
+    List<Field<?>> fields = getCountTypePrefix(countType);
+    fields.add(Fields.stringField(StoreDefinition.AppMetadataStore.NAMESPACE_FIELD, applicationId.getNamespace()));
+    fields.add(Fields.stringField(StoreDefinition.AppMetadataStore.APPLICATION_FIELD, applicationId.getApplication()));
+    fields.add(Fields.stringField(StoreDefinition.AppMetadataStore.VERSION_FIELD, applicationId.getVersion()));
+    return fields;
+  }
+
+  // Do NOT use with type = RunRecordMeta since that needs custom deserialization
```

Comment:
what is the custom deserialization? Can it be added as a type adapter to the Gson?

Reply:
deserializeRunRecordMeta, since we only require deserialization, the type adapter seems more trouble than its worth

Reply:
got it